File: executioncontext/installationcontext/src/test/java/org/jetbrains/bsp/bazel/installationcontext/entities/mappers/InstallationContextJavaPathEntityMapperTest.java
package org.jetbrains.bsp.bazel.installationcontext.entities.mappers;

import io.vavr.control.Option;
import java.nio.file.Paths;
import org.jetbrains.bsp.bazel.executioncontext.api.entries.mappers.ProjectViewToExecutionContextEntityMapperException;
import org.jetbrains.bsp.bazel.installationcontext.entities.InstallationContextJavaPathEntity;
import org.jetbrains.bsp.bazel.projectview.model.ProjectView;
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewJavaPathSection;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class InstallationContextJavaPathEntityMapperTest {

  private InstallationContextJavaPathEntityMapper mapper;

  @BeforeEach
  public void beforeEach() {
    // given
    this.mapper = new InstallationContextJavaPathEntityMapper();
  }

  @Test
  public void shouldReturnSuccessWithJavaPathFromProjectViewIfJavaPathIsSpecifiedInProjectView() {
    // given
    var projectView =
        ProjectView.builder()
            .javaPath(Option.of(new ProjectViewJavaPathSection(Paths.get("/path/to/java"))))
            .build()
            .get();

    // when
    var javaPathTry = mapper.map(projectView);

    // then
    assertTrue(javaPathTry.isSuccess());
    var javaPath = javaPathTry.get();

    var expectedJavaPath = new InstallationContextJavaPathEntity(Paths.get("/path/to/java"));
    assertEquals(expectedJavaPath, javaPath);
  }

  @Test
  public void
      shouldReturnSuccessWithJavaPathFromSystemPropertyIfJavaPathIsNotSpecifiedInProjectView() {
    // given
    var projectView = ProjectView.builder().javaPath(Option.none()).build().get();

    System.setProperty("java.home", "/path/to/java");

    // when
    var javaPathTry = mapper.map(projectView);

    // then
    assertTrue(javaPathTry.isSuccess());
    var javaPath = javaPathTry.get();

    var expectedJavaPath = new InstallationContextJavaPathEntity(Paths.get("/path/to/java"));
    assertEquals(expectedJavaPath, javaPath);
  }

  @Test
  public void
      shouldReturnFailureIfJavaPathIsNotSpecifiedInProjectViewAndItIsImpossibleToObtainItFromSystemPropererties() {
    // given
    var projectView = ProjectView.builder().javaPath(Option.none()).build().get();

    System.clearProperty("java.home");

    // when
    var javaPathTry = mapper.map(projectView);

    // then
    assertTrue(javaPathTry.isFailure());
    assertEquals(
        ProjectViewToExecutionContextEntityMapperException.class,
        javaPathTry.getCause().getClass());

    var expectedMessage =
        "Mapping project view into 'java path' failed! System property 'java.home' is not"
            + " specified.";
    assertEquals(expectedMessage, javaPathTry.getCause().getMessage());
  }
}

File: executioncontext/installationcontext/src/test/java/org/jetbrains/bsp/bazel/installationcontext/entities/mappers/InstallationContextJavaPathEntityMapperTest.kt
package org.jetbrains.bsp.bazel.installationcontext.entities.mappers

import io.kotest.matchers.shouldBe
import org.jetbrains.bsp.bazel.executioncontext.api.entries.mappers.ProjectViewToExecutionContextEntityMapperException
import org.jetbrains.bsp.bazel.installationcontext.entities.InstallationContextJavaPathEntity
import org.jetbrains.bsp.bazel.projectview.model.ProjectView
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewJavaPathSection
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.nio.file.Paths

class InstallationContextJavaPathEntityMapperTest {

    private lateinit var mapper: InstallationContextJavaPathEntityMapper

    @BeforeEach
    fun beforeEach() {
        // given
        this.mapper = InstallationContextJavaPathEntityMapper()
    }

    @Test
    fun `should return success with java path from project view if java path is specified in project view`() {
        // given
        val projectView = ProjectView.Builder(
            javaPath = ProjectViewJavaPathSection(Paths.get("/path/to/java"))
        ).build().get()

        // when
        val javaPathTry = mapper.map(projectView)

        // then
        javaPathTry.isSuccess shouldBe true
        val javaPath = javaPathTry.get()

        val expectedJavaPath = InstallationContextJavaPathEntity(Paths.get("/path/to/java"))
        javaPath shouldBe expectedJavaPath
    }

    @Test
    fun `should return success with java path from system property if java path is not specified in project view`() {
        // given
        val projectView = ProjectView.Builder(javaPath = null).build().get()
        System.setProperty("java.home", "/path/to/java")

        // when
        val javaPathTry = mapper.map(projectView)

        // then
        javaPathTry.isSuccess shouldBe true
        val javaPath = javaPathTry.get()

        val expectedJavaPath = InstallationContextJavaPathEntity(Paths.get("/path/to/java/bin/java"))
        javaPath shouldBe expectedJavaPath
    }

    @Test
    fun `should return failure if java path is not specified in project view and it is not possible to obtain it from system property`() {
        // given
        val projectView = ProjectView.Builder(javaPath = null).build().get()
        System.clearProperty("java.home")

        // when
        val javaPathTry = mapper.map(projectView)

        // then
        javaPathTry.isFailure shouldBe true
        javaPathTry.cause::class shouldBe ProjectViewToExecutionContextEntityMapperException::class
        javaPathTry.cause.message shouldBe "Mapping project view into 'java path' failed! System property 'java.home' is not specified."
    }
}
